### DISCUSSION TITLE
Lower color depths in ICO files

### DISCUSSION THREAD

**Discussion by gonzoMD** [python-pillow/Pillow#6120] [Q&A] [answered]
I am writing a script which converts linux icon themes to Windows icons (Win XP Standard 16,32 and 48px bmp icons).
These icons contain 32bit RGBA, and 8/4bit P images.

Adding the 32bit icons works perfectly, but I can't add the lower color ones. (`inputfile["colordepth"]` contains integer values 32, 8 and 4; `tempsize` is the same value in all three cases)
For testing purposes I just added an empty P mode image (black square), but if I uncomment this code portion, ALL icons are black squares, even the previously appended 32bit ones. 
Is it even possible to add multiple images with different color depth of the same size to an icon?

here is a code snippet:

```python
            if "colordepth" in inputfile:
                    for color in inputfile["colordepth"]:
                        if color == 32:
                            images.append(Image.open(tempfile))
                            sizes.append(tempsize)
                        else:
                            print(color)
                            #im = Image.new(mode="P", size=tempsize)
                            #images.append(im)
                            #sizes.append(tempsize)
                else:
                    images.append(Image.open(tempfile))
                    sizes.append(tempsize)

            im = Image.new(mode="RGBA", size=(48,48))
            im.save(output_file, format="ICO", sizes = sizes, append_images = images, bitmap_format="bmp")
```
Here is the complete source file: https://github.com/gonzoMD/IconThemeConverter/blob/master/IconThemeConverter.py
Here is a reference about what I want: https://docs.microsoft.com/en-us/previous-versions/ms997636(v=msdn.10)?redirectedfrom=MSDN

Thanks in advance

**Comment by nulano** [comment, marked as answer]
> Is it even possible to add multiple images with different color depth of the same size to an icon?

Bitmap saving was added in #5513. Looking at the code, this is not possible.

> Adding the 32bit icons works perfectly, but I can't add the lower color ones.

Looking at the code, the following formats are supported:

https://github.com/python-pillow/Pillow/blob/397a9409959bccd0130d0b51820f2a94e3bf153d/src/PIL/BmpImagePlugin.py#L291-L297

So both 8-bit pallete and 24-bit RGB are supported, but not 4-bit pallete or 16-bit pallete.
Monochrome or 8-bit grayscale are also supported.

These formats are supported when reading:
https://github.com/python-pillow/Pillow/blob/397a9409959bccd0130d0b51820f2a94e3bf153d/src/PIL/BmpImagePlugin.py#L38-L46

If you require support for these other color depths, I would suggest opening an issue with your feature request: https://github.com/python-pillow/Pillow/issues/new/choose

As a side note, this looks like an updated version of the guide you linked: https://docs.microsoft.com/en-us/windows/win32/uxguide/vis-icons

---

Personally, I would not recommend using Pillow to save in ICO format if you wish to strictly adhere to Microsoft documentation. However, PNG or 32-bit BMP seems to work for most modern uses (e.g. favicons).

**Reply by gonzoMD** [reply]
Thanks for your reply.

further tests by me, when loading my resulting icons in Greenfish Icon Editor, show that there are only 16 colors in a paletted image, when I just create a solid filled image. So the colors seem to be not a problem. 
So iiuc it is possible to add several eg. 32x32px icons, but no different ones and always the latest appended one is used for this size.

OK, so I will add an issue and hope that someone is interested in implementing it.

The old version of the MS documentation was intended, because I want my resulting icons to be ReactOS compatible, which uses XP icon format (no PNG, max 48x48). 

Yeah. Dealing with an almost dead file format, or at least an old version of a further evolved one can be PITA because of missing use cases. But there are hundreds of icons to convert and thinking about to convert them by hand makes my cringe.

Another question: How can I set eg. Magenta #ff00ff as transparent color?

**Reply by nulano** [reply]
> further tests by me, when loading my resulting icons in Greenfish Icon Editor, show that there are only 16 colors in a paletted image, when I just create a solid filled image. So the colors seem to be not a problem.

Not entirely sure what you mean by this. From the code I linked above, Pillow will always create an 8-bit palette, no matter how many entries are actually used.

> Another question: How can I set eg. Magenta #ff00ff as transparent color?

I don't think this is possible. I can't find any reference to this in the ICO format specification: https://docs.microsoft.com/en-us/previous-versions/ms997538(v=msdn.10)?redirectedfrom=MSDN
It is probably just a feature your image editor uses to generate transparency information.
I think you can set a transparent color by using an RGBA palette with a transparent color.

The other way of specifying transparent pixels is via a 1bpp AND mask image (see specification).
This is currently not supported when saving in Pillow (a fully opaque mask is always saved).

**Reply by nulano** [reply]
> OK, so I will add an issue and hope that someone is interested in implementing it.

Just to link to it, the issue is #6121.

**Reply by radarhere** [reply]
The issue was resolved by #6122
